cpp/src/arrow/adapters/orc/adapter.cc

```diff
@@ -0,0 +1,698 @@
+
+  uint64_t NumberOfRows() { return reader_->getNumberOfRows(); }
+
+  Status ReadSchema(std::shared_ptr<Schema>* out) {
+    const liborc::Type& type = reader_->getType();
+    return type_to_schema(type, out);
+  }
+
+  Status type_to_schema(const liborc::Type& type, std::shared_ptr<Schema>* out) {
+    if (type.getKind() != liborc::STRUCT) {
+      return Status::NotImplemented(
+          "Only ORC files with a top-level struct "
+          "can be handled");
+    }
+    auto size = type.getSubtypeCount();
+    auto fields = std::vector<std::shared_ptr<Field>>();
+    for (uint64_t child = 0; child < size; ++child) {
+      std::shared_ptr<DataType> elemtype;
+      RETURN_NOT_OK(get_dtype(type.getSubtype(child), &elemtype));
+      std::string name = type.getFieldName(child);
+      fields.push_back(field(name, elemtype));
+    }
+    std::list<std::string> keys = reader_->getMetadataKeys();
+    std::shared_ptr<KeyValueMetadata> metadata = nullptr;
+    if (!keys.empty()) {
+      metadata = std::make_shared<KeyValueMetadata>();
+      for (auto it = keys.begin(); it != keys.end(); ++it) {
+        metadata->Append(*it, reader_->getMetadataValue(*it));
+      }
+    }
+
+    *out = std::make_shared<Schema>(fields, metadata);
+    return Status::OK();
+  }
+
+  Status Read(std::shared_ptr<RecordBatch>* out) {
+    liborc::RowReaderOptions opts;
+    return read_batch(opts, NumberOfRows(), out);
+  }
+
+  Status Read(const std::list<uint64_t>& include_indices,
+              std::shared_ptr<RecordBatch>* out) {
+    liborc::RowReaderOptions opts;
+    opts.includeTypes(include_indices);
+    return read_batch(opts, NumberOfRows(), out);
+  }
+
+  Status ReadStripe(uint64_t stripe, std::shared_ptr<RecordBatch>* out) {
+    liborc::RowReaderOptions opts;
+    RETURN_NOT_OK(select_stripe(&opts, stripe));
+    return read_batch(opts, stripes_[stripe].num_rows, out);
+  }
+
+  Status ReadStripe(uint64_t stripe, const std::list<uint64_t>& include_indices,
+                    std::shared_ptr<RecordBatch>* out) {
+    liborc::RowReaderOptions opts;
+    RETURN_NOT_OK(select_stripe(&opts, stripe));
+    opts.includeTypes(include_indices);
+    return read_batch(opts, stripes_[stripe].num_rows, out);
+  }
+
+  Status select_stripe(liborc::RowReaderOptions* opts, uint64_t stripe) {
+    if (stripe >= stripes_.size()) {
+      std::stringstream ss;
+      ss << "Out of bounds stripe: " << stripe;
+      return Status::Invalid(ss.str());
+    }
+    opts->range(stripes_[stripe].offset, stripes_[stripe].length);
+    return Status::OK();
+  }
+
+  Status read_batch(const liborc::RowReaderOptions& opts, uint64_t nrows,
+                    std::shared_ptr<RecordBatch>* out) {
+    std::unique_ptr<liborc::RowReader> rowreader;
+    std::unique_ptr<liborc::ColumnVectorBatch> batch;
+    try {
+      rowreader = reader_->createRowReader(opts);
+      batch = rowreader->createRowBatch(std::min(nrows, (uint64_t)1000));
```

Comment:
I'm not terribly happy with this magic number here. Since data has to be copied out of the orc reader and into the arrow structure, it doesn't make sense to allocate a batch the same size as the output (doubling memory usage), but you also don't want a really tiny batch. This could be configurable if needed, but I'd rather have users avoid this detail. `liborc` does expose a memory estimate for a read, so the fixed number of rows could be switched to a fixed memory size, and the batch size computed from there.

Reply:
It would be nice if you refactor this out as a constant declared somewhere, e.g. `constexpr uint64_t kReadRowsBatch = 1000`. One approach we used in parquet-cpp was to implement a specialization in the parquet-cpp code that could directly read into pre-allocated memory. I don't expect this to be implemented in this PR but it could be a worthwhile optimisation if this here proves to be too slow (we had a 30% performance improvement by that change in parquet_arrow).

Reply:
> One approach we used in parquet-cpp was to implement a specialization in the parquet-cpp code that could directly read into pre-allocated memory.

I'd also expect substantial performance improvements with this route (could avoid the extra allocations, all the casting, dropping of dictionary encoding, etc...). This would require the arrow implementation move to liborc instead (an option that I mentioned above), as none of the required interfaces are exposed in the `liborc` library. I don't think this would be a problem to do later, as the user-api exposed here shouldn't change, just the build process.